src/app/styles/bootstrap/_root.scss

```diff
@@ -0,0 +1,79 @@
+@use "sass:color";
+
+@function contrast($lightness, $threshold) {
+  @return hsl(0, 0%, calc((#{$lightness} - #{$threshold}) * -100));
+}
+
+@function lighten($lightness, $offset) {
+  @return calc(#{$lightness} + #{$offset});
+}
+
+@function darken($lightness, $offset) {
+  @return calc(#{$lightness} - #{$offset});
+}
+
+@mixin create-variant($property, $variant, $lightness) {
+  --baw-#{$property}#{$variant}: hsl(
+    var(--baw-#{$property}-hue),
+    var(--baw-#{$property}-saturation),
+    #{$lightness}
+  );
+  --baw-#{$property}#{$variant}-contrast: #{contrast(
+      #{$lightness},
+      var(--baw-contrast-threshold)
+    )};
+}
+
+@mixin create-variable(
+  $property,
+  $hue,
+  $saturation,
+  $lightness: 50%,
+  $lighter-offset: 10%,
+  $lightest-offset: 25%,
+  $darker-offset: 10%,
+  $darkest-offset: 25%
+) {
+  --baw-#{$property}-hue: #{$hue};
+  --baw-#{$property}-saturation: #{$saturation};
+  --baw-#{$property}-lightness: #{$lightness};
+
+  @include create-variant($property, "", var(--baw-#{$property}-lightness));
+
+  @include create-variant(
+    $property,
+    "-lighter",
+    lighten(var(--baw-#{$property}-lightness), #{$lighter-offset})
+  );
+
+  @include create-variant(
+    $property,
+    "-lightest",
+    lighten(var(--baw-#{$property}-lightness), #{$lightest-offset})
+  );
+
+  @include create-variant(
+    $property,
+    "-darker",
+    darken(var(--baw-#{$property}-lightness), #{$darker-offset})
+  );
+
+  @include create-variant(
+    $property,
+    "-darkest",
+    darken(var(--baw-#{$property}-lightness), #{$darkest-offset})
+  );
+}
+
+:root {
+  --baw-contrast-threshold: 50%;
+
+  @include create-variable("primary", 216, 98%, 52%);
+  @include create-variable("secondary", 208, 7%, 46%);
+  @include create-variable("success", 152, 69%, 31%);
+  @include create-variable("info", 190, 90%, 50%);
+  @include create-variable("warning", 45, 100%, 51%);
+  @include create-variable("danger", 354, 70%, 54%);
+  @include create-variable("light", 210, 17%, 98%);
+  @include create-variable("dark", 210, 11%, 15%);
```

Comment:
are these all bootstrap defaults? 

maybe we should referencing their original variables?

Reply:
They are bootstrap defaults converted to hsl, AFAIK there is no easy way to convert the rgb values of bootstrap to hsl. I have however just discovered a potential way of handling all of the colour calculations in rgb however if we wanted to use bootstrap values as defaults. I ruled it out as a possibility as you would need to set transparency values on the colours (which would be bad on things like image backgrounds), however it appears i could use the following solution to bypass that: https://gist.github.com/tqc/2564280. Colour contrast calculations may be slightly better/worse as well